src/backend/distributed/commands/call.c

```diff
@@ -0,0 +1,204 @@
+/*-------------------------------------------------------------------------
+ *
+ * call.c
+ *    Commands for call remote stored procedures.
+ *
+ * Copyright (c) 2019, Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#if PG_VERSION_NUM >= 110000
+
+#include "catalog/pg_proc.h"
+#include "commands/defrem.h"
+#include "distributed/colocation_utils.h"
+#include "distributed/commands.h"
+#include "distributed/connection_management.h"
+#include "distributed/master_metadata_utility.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/remote_commands.h"
+#include "distributed/shard_pruning.h"
+#include "distributed/version_compat.h"
+#include "distributed/worker_manager.h"
+#include "nodes/nodeFuncs.h"
+#include "nodes/parsenodes.h"
+#include "nodes/primnodes.h"
+#include "tcop/dest.h"
+#include "utils/lsyscache.h"
+#include "utils/syscache.h"
+
+static void PipePgResultToDestReceiver(TupleDesc resultDesc, PGresult *result,
+									   DestReceiver *dest);
+static bool CallFuncExprRemotely(CallStmt *callStmt,
+								 DistObjectCacheEntry *procedure,
+								 FuncExpr *funcExpr, const char *queryString,
+								 DestReceiver *dest);
+
+/*
+ * CallDistributedProcedure calls a stored procedure on the worker if possible.
+ */
+bool
+CallDistributedProcedureRemotely(CallStmt *callStmt, const char *queryString,
+								 DestReceiver *dest)
+{
+	DistObjectCacheEntry *procedure = NULL;
+	FuncExpr *funcExpr = callStmt->funcexpr;
+	Oid functionId = funcExpr->funcid;
+
+	procedure = LookupDistObjectCacheEntry(ProcedureRelationId, functionId, 0);
+	if (procedure == NULL)
+	{
+		return false;
+	}
+
+	return CallFuncExprRemotely(callStmt, procedure, funcExpr, queryString, dest);
+}
+
+
+/*
+ * PipePgResultToDestReceiver outputs a PGresult into a DestReceiver
+ * We assume DestReceiver expects a TTSOpsVirtual tuple slot
+ */
+static void
+PipePgResultToDestReceiver(TupleDesc resultDesc, PGresult *result, DestReceiver *dest)
```

Comment:
Alternative approach is https://github.com/citusdata/citus/pull/2998/, just in case.